_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,340 @@
+> It was fixed with by forcing the producer to send the consumer group metadata along with the offsets to commit ([KIP-447](https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics)).
+
+To guarantee message ordering, a given producer can have at most one open transaction (they are executed serially).
+Consumers with `isolation.level=read_committed` are not allowed to advance to offsets which include open transactions, while messages from aborted transactions are filtered out.
+
+The transaction coordinator is a module running inside every Kafka broker.
+Each coordinator owns a subset of the partitions in the internal `__transaction_state` topic (i.e. the partitions for which its broker is the leader).
+The coordinator automatically aborts any ongoing transaction that is not completed within `transaction.timeout.ms`.
+
+It is important to note that transactions are only supported inside a single Kafka cluster.
+If the transaction scope includes processors with externally observable side-effects, you would need to use an additional component such as `ChainedKafkaTransactionManager` from the SpringBoot project.
+This component is able to synchronize a database transaction with the Kafka transaction.
+
+#### Transaction metadata
+
+A control record is a special type of record that is used for internal operations within Kafka, rather than for normal message processing.
+They are primarily used by the Kafka brokers themselves to manage topics, partitions, and other metadata, but may sometimes be used by administrators or monitoring tools as well.
+Control records are typically not visible to Kafka consumers, as they are meant to be processed only by the Kafka brokers.
+
+An end transaction marker (`COMMIT` or `ABORT`) is a special control record that indicates the end of a transaction (decision).
+It is used to ensure that the transaction is either completed successfully or aborted in case of a failure.
+The end transaction marker is added to the log after all the records related to the transaction have been written to the log.
+
+Non-transactional records are considered decided immediately, but transactional records are only decided when the corresponding commit or abort marker is written.
+
+A transaction goes through the following states:
+
+1. **Undecided** (Ongoing)
+2. **Decided and unreplicated** (PrepareCommit\|PrepareAbort)
+3. **Decided and replicated** (CompleteCommit\|CompleteAbort)
+
+This is handled through a process similar to the two-phase commit protocol.
+Every time some data is sent to a new partition inside a transaction, we have a new `Ongoing` control record.
+When the transaction completes, we have the `PrepareCommit` or `PrepareAbort` control record and the transaction coordinator is doomed to commit or abort.
+The final step consists of writing the end transaction markers to each involved partition, and the `CompleteCommit` or `CompleteAbort` control record.
```

Comment:
This could be improved by explaining that these states are from the transaction coordinator's PoV. Right now you're blurring commit markers in user partitions with the (persistent) state in the coordinator, and that will be confusing to anyone who doesn't know this stuff already.

Reply:
You are right. That could be confusing, so let me add a line here.